### PR TITLE
Assignment with booleans + follow-up on random_uniform for GLn and hypersphere

### PR DESCRIPTION
**Pull request by pchauchat**
This PR allows `gs.assignment` to work with booleans (and indices given as tensors) in tf.

It is also a follow-up on #583

### CODE REVIEW

**Comment by pchauchat** [comment]
`gs.assignment` was heavily refactored to take into account arrays and booleans. For numpy and pytorch, I tried to take as much advantage as possible of their slicing opportunities, to keep the cost low (although I don't guarantee anything).

I defined more general check functions, to see if an object is some kind of boolean (mostly a native one or an array of bools), and if it is iterable (list, tuple, array of dim > 0). Maybe it will be useful for other functions too.

**Comment by pchauchat** [comment]
`gs.assignment` was heavily refactored to take into account arrays and booleans. For numpy and pytorch, I tried to take as much advantage as possible of their slicing opportunities, to keep the cost low (although I don't guarantee anything).

I defined more general check functions, to see if an object is some kind of boolean (mostly a native one or an array of bools), and if it is iterable (list, tuple, array of dim > 0). Maybe it will be useful for other functions too.

File: geomstats/_backend/numpy/__init__.py [hunk 1]
```diff
@@ -118,6 +118,24 @@ def get_mask_i_float(i, n):
     return mask_i_float
 
 
+def _is_boolean(x):
+    if isinstance(x, bool):
+        return True
+    if isinstance(x, (tuple, list)):
+        return _is_boolean(x[0])
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:125]: Why are you only testing the 0th element of the tuple?

File: geomstats/_backend/numpy/__init__.py [hunk 2]
```diff
@@ -146,19 +164,20 @@ def assignment(x, values, indices, axis=0):
...
-            index = (index,)
-        if len(index) < len(shape(x)):
-            for n_axis in range(shape(x)[axis]):
-                extended_index = index[:axis] + (n_axis,) + index[axis:]
-                x_new[extended_index] = values[nb_index]
-        else:
-            x_new[index] = values[nb_index]
+
+    use_vectorization = hasattr(indices, '__len__') and len(indices) < ndim(x)
+    if _is_boolean(x):
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:169]: Should it be the indices that are tested to be boolean or not?

File: geomstats/_backend/numpy/__init__.py [hunk 3]
```diff
@@ -146,19 +164,20 @@ def assignment(x, values, indices, axis=0):
...
+        x_new[indices] = values
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:170]: This is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 4]
```diff
@@ -303,10 +321,16 @@ def assignment_by_sum(x, values, indices, axis=0):
...
     """
-    if not isinstance(values, list):
+    if _is_boolean(indices):
+        indices = [index for index, val in enumerate(indices) if val]
+
+    if tf.rank(values) == 0:
         return _assignment_single_value_by_sum(x, values, indices, axis)
 
-    if not isinstance(indices, list):
+    single_index = not isinstance(indices, list)
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:330]: These 5 lines are not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 5]
```diff
@@ -258,11 +268,19 @@ def _assignment_single_value_by_sum(x, value, indices, axis=0):
         Copy of x where value was added at all indices (and possibly along
         an axis).
     """
+    if _is_boolean(indices):
+        indices = [index for index, val in enumerate(indices) if val]
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:272]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 6]
```diff
@@ -258,11 +268,19 @@ def _assignment_single_value_by_sum(x, value, indices, axis=0):
...
+
     single_index = not isinstance(indices, list)
+    if tf.is_tensor(indices):
+        single_index = ndim(indices) <= 1 and sum(indices.shape) <= ndim(x)
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:276]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 7]
```diff
@@ -258,11 +268,19 @@ def _assignment_single_value_by_sum(x, value, indices, axis=0):
...
     if single_index:
         indices = [indices]
+
     if isinstance(indices[0], tuple):
         use_vectorization = (len(indices[0]) < ndim(x))
+    elif tf.is_tensor(indices[0]) and ndim(indices[0]) >= 1:
+        use_vectorization = (len(indices[0]) < ndim(x))
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:283]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 8]
```diff
@@ -303,10 +321,16 @@ def assignment_by_sum(x, values, indices, axis=0):
     If a single value is provided, it is assigned at all the indices.
     If a list is given, it must have the same length as indices.
     """
-    if not isinstance(values, list):
+    if _is_boolean(indices):
+        indices = [index for index, val in enumerate(indices) if val]
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:325]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 9]
```diff
@@ -134,6 +134,16 @@ def any(x, axis=None):
     return tf.math.reduce_any(tf.cast(x, bool), axis=axis)
 
 
+def _is_boolean(x):
+    if isinstance(x, bool):
+        return True
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:139]: This if is not tested.

File: geomstats/_backend/pytorch/__init__.py [hunk 10]
```diff
@@ -599,20 +617,20 @@ def assignment_by_sum(x, values, indices, axis=0):
...
-        if len(index) < len(shape(x)):
-            for n_axis in range(shape(x)[axis]):
-                extended_index = index[:axis] + (n_axis,) + index[axis:]
-                x_new[extended_index] += values[nb_index]
-        else:
-            x_new[index] += values[nb_index]
+
+    use_vectorization = hasattr(indices, '__len__') and len(indices) < ndim(x)
+    if _is_boolean(x):
+        x_new[indices] = values
```
Comment (ninamiolane) [inline geomstats/_backend/pytorch/__init__.py:623]: This if is not tested.

File: geomstats/_backend/pytorch/__init__.py [hunk 11]
```diff
@@ -554,20 +572,20 @@ def assignment(x, values, indices, axis=0):
...
-        if len(index) < len(shape(x)):
-            for n_axis in range(shape(x)[axis]):
-                extended_index = index[:axis] + (n_axis,) + index[axis:]
-                x_new[extended_index] = values[nb_index]
-        else:
-            x_new[index] = values[nb_index]
+
+    use_vectorization = hasattr(indices, '__len__') and len(indices) < ndim(x)
+    if _is_boolean(x):
+        x_new[indices] = values
```
Comment (ninamiolane) [inline geomstats/_backend/pytorch/__init__.py:578]: This if is not tested.

File: geomstats/_backend/pytorch/__init__.py [hunk 12]
```diff
@@ -526,6 +526,24 @@ def get_mask_i_float(i, n):
     return mask_i_float
 
 
+def _is_boolean(x):
+    if isinstance(x, bool):
+        return True
```
Comment (ninamiolane) [inline geomstats/_backend/pytorch/__init__.py:531]: This if is not tested.

File: geomstats/_backend/pytorch/__init__.py [hunk 13]
```diff
@@ -526,6 +526,24 @@ def get_mask_i_float(i, n):
...
+    if isinstance(x, (tuple, list)):
+        return _is_boolean(x[0])
```
Comment (ninamiolane) [inline geomstats/_backend/pytorch/__init__.py:533]: This if is not tested.

File: geomstats/_backend/pytorch/__init__.py [hunk 14]
```diff
@@ -526,6 +526,24 @@ def get_mask_i_float(i, n):
...
+    if torch.is_tensor(x):
+        return x.dtype in [torch.bool, torch.uint8]
+    return False
```
Comment (ninamiolane) [inline geomstats/_backend/pytorch/__init__.py:536]: This is not tested.

File: geomstats/_backend/numpy/__init__.py [hunk 15]
```diff
@@ -190,19 +209,20 @@ def assignment_by_sum(x, values, indices, axis=0):
...
-        if len(index) < len(shape(x)):
-            for n_axis in range(shape(x)[axis]):
-                extended_index = index[:axis] + (n_axis,) + index[axis:]
-                x_new[extended_index] += values[nb_index]
-        else:
-            x_new[index] += values[nb_index]
+
+    use_vectorization = hasattr(indices, '__len__') and len(indices) < ndim(x)
+    if _is_boolean(x):
+        x_new[indices] = values
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:215]: This if is not tested.

File: geomstats/_backend/numpy/__init__.py [hunk 16]
```diff
@@ -118,6 +118,24 @@ def get_mask_i_float(i, n):
     return mask_i_float
 
 
+def _is_boolean(x):
+    if isinstance(x, bool):
+        return True
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:123]: This if is not tested.

File: geomstats/_backend/numpy/__init__.py (hunk 1 above)
Comment (ninamiolane) [reply geomstats/_backend/numpy/__init__.py:125]: This if is not tested.

File: geomstats/_backend/numpy/__init__.py [hunk 17]
```diff
@@ -118,6 +118,24 @@ def get_mask_i_float(i, n):
...
+    if isinstance(x, (tuple, list)):
+        return _is_boolean(x[0])
+    if isinstance(x, np.ndarray):
+        return x.dtype == bool
+    return False
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:128]: This is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 18]
```diff
@@ -341,11 +365,19 @@ def _assignment_single_value(x, value, indices, axis=0):
         Copy of x where value was assigned at all indices (and possibly
         along an axis).
     """
+    if _is_boolean(indices):
+        indices = [index for index, val in enumerate(indices) if val]
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:369]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 19]
```diff
@@ -341,11 +365,19 @@ def _assignment_single_value(x, value, indices, axis=0):
...
+
     single_index = not isinstance(indices, list)
+    if tf.is_tensor(indices):
+        single_index = ndim(indices) <= 1 and sum(indices.shape) <= ndim(x)
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:373]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 20]
```diff
@@ -341,11 +365,19 @@ def _assignment_single_value(x, value, indices, axis=0):
...
     if single_index:
         indices = [indices]
+
     if isinstance(indices[0], tuple):
         use_vectorization = (len(indices[0]) < ndim(x))
+    elif tf.is_tensor(indices[0]) and ndim(indices[0]) >= 1:
+        use_vectorization = (len(indices[0]) < ndim(x))
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:380]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 21]
```diff
@@ -388,10 +420,16 @@ def assignment(x, values, indices, axis=0):
...
+    if _is_boolean(indices):
+        indices = [index for index, val in enumerate(indices) if val]
+
+    if tf.rank(values) == 0:
         return _assignment_single_value(x, values, indices, axis)
 
-    if not isinstance(indices, list):
+    single_index = not isinstance(indices, list)
+    if tf.is_tensor(indices):
+        single_index = ndim(indices) <= 1 and sum(indices.shape) <= ndim(x)
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:431]: This if is not tested.

File: geomstats/_backend/tensorflow/__init__.py [hunk 22]
```diff
@@ -388,10 +420,16 @@ def assignment(x, values, indices, axis=0):
...
+    if single_index:
         indices = [indices]
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:433]: This if is not tested.

File: geomstats/_backend/numpy/__init__.py (hunk 1 above)
Comment (pchauchat) [reply geomstats/_backend/numpy/__init__.py:125]: I made the assumption that either all elements are booleans or no one is. I don't think it makes sense to make the code heavier and longer to run to be able to account for cases where booleans would be mixed with non-boolean values. I am not even sure Numpy handles them.

File: geomstats/_backend/numpy/__init__.py (hunk 2 above)
Comment (pchauchat) [reply geomstats/_backend/numpy/__init__.py:169]: Good catch
